[Storybook]: Typecheck in another process for non-CI builds

fallaciousreasoning opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/36913

The speedup is pretty drastic - 1.8 minutes to 16 seconds ~6x faster

Followup for https://github.com/brave/brave-core/pull/22658

Additionally, this PR converts the Storybook files from CommonJS to Typescript ESModules, which should help catch errors while working on Storybook (since V7 Storybook supports its config being in Typescript).

Security/Privacy review: 
https://github.com/brave/reviews/issues/1556

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-arm64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-x64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

petemill commented on .storybook/webpack.config.js:
```diff
@@ -148,11 +149,12 @@ module.exports = async ({ config, mode }) => {
       test: /\.(ts|tsx)$/,
       loader: 'ts-loader',
       options: {
+        transpileOnly: !isCI,
```
How about we just specify this via a cli / environment variable? It would solve two issues:
1. Different ways to detect different CI environments other than the jenkins-specific vars in `ci.ts`
2. Being able to run with checking on local builds when trying to fix errors

mherrmann replied:
I agree with @petemill. This should be explicit. My preferred approach (if feasible) would be via a separate command that does nothing else but run the type checks, similarly to `npm run presubmit`. Otherwise, via a CLI argument. If that is not possible either, then via an environment variable whose sole purpose is to turn the feature on or off. For example, `TRANSPILE_ONLY`. I can make the necessary changes in CI.

I also want to explain why I have the above preference. In short, it is because explicit is better than implicit. There are several ways in which the current implementation can cause trouble, just because it is implicit:

1. We are planning to switch from Jenkins to TeamCity. Once that happens, the code proposed here will no longer work. But we won't even notice! Failing type checks will simply not show up as errors in CI. This is because the current code encodes knowledge about Jenkins. It's really not something it should do.
2. Like @petemill said, what if I do want to run the check synchronously locally? Then I have to fake Jenkins' environment variables on my local machine. That's not just awkward but also counter-intuitive.

I was personally affected by a problem caused by an implicit implementation in related code in https://github.com/brave/brave-browser/issues/34984. It was confusing and cost me time to debug. I believe that every implicit solution can be turned into an explicit one, maybe with a very small amount of additional manual one-time effort on the part of the developer. But it avoids painful and unnecessary debugging sessions.

I like the Zen of Python, which also talks about explicitness:

```
Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.[c]
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than right now.[d]
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea – let's do more of those!
```

petemill commented:
> My preferred approach (if feasible) would be via a separate command that does nothing else but run the type checks, similarly to npm run presubmit

I don't think we need to go that far. Let's remember that storybook is primarily an additive tool for local devs. Yes, CI does provide an export of this tool, and we want to make sure that passes. But the actual frontend code that is used by the browser is still type checked within the regular browser build. This change doesn't affect that.

fallaciousreasoning commented:
Okay, so I don't think its a good idea to separate typechecking from the Storybook command - I can't imagine a scenario where you'd want to run Storybook without doing the typechecks.

I would like to end up in the following situation:
1. When you run storybook locally, by default, it runs using `transpileOnly` and ForkTsChecker.
2. On CI it runs the `ts-loader` typechecker, so if the typecheck fails, the Storybook step fails.

If you're running a `watch` style build, you should be using `ForkTsChecker`. I'm not opposed to introducing a flag/env variable for doing this, but that should be the default.

What about this:
1. When you run `npm run storybook` it uses `ForkTsChecker`. This command is only used for development
2. When you run `npm run build-storybook` we use the typechecker built in to `ts-loader`. This command generates a static Storybook deployment and it **should** fail if there are problems with the types.

We can also add a custom flag/environment variable for configuring this, but I'm not really sure why you'd want to - output in dev mode is the same as what we have currently, and prod builds won't change.

fallaciousreasoning commented:
@petemill @mherrmann I've updated this PR with a `DONT_FORK_TS_CHECKER` environment variable which forces this feature off, and changed things so it only runs in `dev` mode (which is `npm run storybook`). `npm run build-storybook` will continue to use the existing (non-forked) behavior. We don't provide a way to opt in to the forking for a static build, because that command should fail when typechecking fails.

mherrmann commented on .storybook/options.ts:
```diff
@@ -0,0 +1,16 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// If this env variable is set, we won't fork the type checker.
+const DONT_FORK_TS_CHECKER = 'DONT_FORK_TS_CHECKER'
```
Do we know that we are going to use this environment variable, or is it here just in case? If the latter, then I think we should remove the environment variable and add it in the future if we find out that we really need it.

fallaciousreasoning replied:
Sweet, I only added it because I thought you wanted an explicit way of changing back. More than happy to remove it :smile:

mherrmann commented on .storybook/webpack.config.ts:
```diff
@@ -115,8 +115,9 @@ function useMockedModules(moduleNames) {
-module.exports = async ({ config, mode }) => {
-  const isDevMode = mode === 'development'
+export default async ({ config, mode }) => {
+  const isDevMode = mode.toLowerCase() === 'development'
```
It's funny that the mode seems to be `development`, but the entry in `argv` is `dev`.

fallaciousreasoning replied:
So the entry in `argv` is `dev` because the command which is executed is `{node} path/to/storybook.js {build|dev}`. The mode there is the separate `Webpack` mode, (I think it's coming from the `NODE_ENV` environment variable) which is either `DEVELOPMENT` or `PRODUCTION` (it's changed based on whether Storybook is doing a static build or a dev build).

`config.mode` is `development` | `production`, so  maybe we should be using that instead - this was just fixing a bug I noticed that `isDevMode` was always false.

mherrmann reviewed (changes requested):
This new implementation is much better in my opinion. I would still remove `DONT_FORK_TS_CHECKER` unless we know that we are going to need it.

fallaciousreasoning commented:
@mherrmann this has been updated on top of the changes from https://github.com/brave/brave-core/pull/21443

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/sb-partial/88317221ddbb44db582e058e49423d845400e6cb/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/sb-partial/a4d4ffef713e3dcc3a1f6540324e1f8e1d48f4f6/storybook/index.html)

petemill commented on .storybook/options.ts:
```diff
@@ -0,0 +1,10 @@
+// there's no disadvantage to doing so. Static builds use the ts-loader
+// typechecker, so the build fails when typechecking fails.
+// Note: argv will include 'dev' for dev builds and 'build' for static builds.
+export const forkTsChecker = process.argv.includes('dev')
```
This _could_ be a little flakey if Storybook changes how their internal commands are constructed?

fallaciousreasoning replied:
Agreed, it feels a little icky to depend on but there doesn't seem to be anything else to let us tell the commands apart.

petemill reviewed (approved):
This is a good compromise and will help a lot of devs out. If a developer needs to enable the stricter build locally then they can still run `npm run build-storybook` / `storybook build`, or we could add a param when needed. It's (very) slightly flakey because the `options.ts` value is determined on the structure of the storybook command (which in the current storybook version is either `storybook dev` or `storybook build`), but I think that's an ok trade-off as we should know pretty quickly if that's broken. And surely they won't change their command structure (again...?)...

fallaciousreasoning commented:
*Note:* audit failure is unrelated to this PR

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/sb-partial/eed9c8090c16d851b7c7389a9cb35733cbf520a3/storybook/index.html)
